### ISSUE TITLE
IPv4 padding becomes random after a few packets

### ISSUE THREAD

**Issue by cha0sCat** [pouriyajamshidi/flat#30]
When using the original code, after the program outputs about 150 lines of connection logs at the beginning, there is no more output.

![Snipaste_2024-04-30_10-39-31](https://github.com/pouriyajamshidi/flat/assets/47235156/b4357057-3a11-42d3-b3c7-474c43b91b98)

So, I add some logs to it
https://github.com/cha0sCat/flat/blob/a0010830e35163d96aa2106a12003ae67538b656/internal/probe/probe.go#L233
https://github.com/cha0sCat/flat/blob/a0010830e35163d96aa2106a12003ae67538b656/internal/probe/probe.go#L242

Then I found that the IP prefix becomes random after a few seconds of the program running.
(This is an ipv4 connection actually)
```
2024/04/30 01:48:13 Received packet: [177 245 45 248 189 87 2 0 48 0 255 255 10 10 34 52 243 232 172 230 189 87 2 0 48 0 255 255 10 10 32 37 56 225 46 37 6 64 1 1 202 69 106 7 190 87 2 0]
2024/04/30 01:48:13 Received packetAttrs: {SrcIP:b1f5:2df8:bd57:200:3000:ffff:a0a:2234 DstIP:f3e8:ace6:bd57:200:3000:ffff:a0a:2025 SrcPort:14561 DstPort:11813 Protocol:6 TTL:64 Syn:true Ack:true TimeStamp:659423633229258}
```

please notice that we can still find the actual ipv4 address in the raw packet:
177 245 45 248 189 87 2 0 48 0 255 255 **10 10 34 52** 243 232 172 230 189 87 2 0 48 0 255 255 **10 10 32 37** 56 225 46 37 6 64 1 1 202 69 106 7 190 87 2 0

comparing to normal packets:
```
2024/04/30 01:53:55 Received packet: [0 0 0 0 0 0 0 0 0 0 255 255 10 10 38 52 0 0 0 0 0 0 0 0 0 0 255 255 10 10 32 37 56 227 186 59 17 64 0 0 190 64 22 190 13 88 2 0]
2024/04/30 01:53:55 Received packetAttrs: {SrcIP:::ffff:10.10.38.52 DstIP:::ffff:10.10.32.37 SrcPort:14563 DstPort:47675 Protocol:17 TTL:64 Syn:false Ack:false TimeStamp:659766000369854}
(UDP) | src: 10.10.32.37:47675          dst: 10.10.38.52:14563          TTL: 64         latency: 0.062 ms
2024/04/30 01:53:55 Received packet: [0 0 0 0 0 0 0 0 0 0 255 255 10 10 32 53 0 0 0 0 0 0 0 0 0 0 255 255 10 10 32 37 2 249 128 171 6 64 0 0 73 202 22 190 13 88 2 0]
```

**Comment by cha0sCat** [comment]
PR sent #31 

I'm very new to C so please bear with me if the code looks bad :P

**Comment by pouriyajamshidi** [comment]
Hey @cha0sCat,

Thanks for debugging this and the detailed explanation.

Please give me some time to review your PR.

**Comment by pouriyajamshidi** [comment]
@cha0sCat, I am not able to reproduce this. I use both IPv4 and IPv6, running the program for hours but the IPv4 never starts with anything but `SrcIP:::ffff:x.x.x.x`.

1. Are you using the latest commit?
2. What kernel version are you using?

**Comment by cha0sCat** [comment]
For C, reading from uninitialized var is an undefined behaviour, so the behaviour might be quite different on different OS and compilers. (https://www.learncpp.com/cpp-tutorial/uninitialized-variables-and-undefined-behavior/)

1. Yes, I'm using the latest commit
2. For building, I use GitHub actions with the ubuntu-latest image; I'm not sure what the exact kernel version is.
Here's the binary file: https://github.com/cha0sCat/flat/actions/runs/8887259908
(you can also use this compiled version to see if it can reproduce)

For running I use `Linux HOSTNAMEHERE 6.5.11-4-pve #1 SMP PREEMPT_DYNAMIC PMX 6.5.11-4 (2023-11-20T10:19Z) x86_64 x86_64 x86_64 GNU/Linux` with ubuntu 20.04


Some additional context:
1. I run flat on a heavy network load VM with about 200MByte/s in & out. (So it might be reusing the same block of memory very frequently)
2. To build on ubuntu-latest and run on ubuntu-20.04 without any glibc version issue, I mod the build command to make a fully static build. For the full workflow file, please check: https://github.com/cha0sCat/flat/actions/runs/8887259908/workflow

**Comment by pouriyajamshidi** [comment]
@cha0sCat so, I added a static release here:

https://github.com/pouriyajamshidi/flat/releases/download/v0.2.0/flat

It is based on a change here:

https://github.com/pouriyajamshidi/flat/blob/c6f34aedff6bc77259524d51ab86681730f3209b/bpf/flat.c#L146

Please give it a shot and let me know if it solves it.

**Comment by cha0sCat** [comment]
No, the issue still exists; I'm using the exact same VM as the first time.

Also, I want to correct my previous report. This issue happens not a few seconds after running but almost less than one second.


![output](https://github.com/pouriyajamshidi/flat/assets/47235156/5473379e-a645-42ab-a444-e5436a98bd60)

**Comment by pouriyajamshidi** [comment]
ah, that is a shame. I was hoping that the `NULL` assignment would fix it.

I understand that your PR fixes it but I was hoping to fix it more fundamentally (ensuring other `packet_t` fields are not affected as well) since we have a good test case here.

Could you please also give the latest release one more shot? The link:

https://github.com/pouriyajamshidi/flat/releases/tag/v0.2.1

The change:

https://github.com/pouriyajamshidi/flat/blob/9f12e8484452c7643cff0b82a6acea3563690584/bpf/flat.c#L155

Once again, thanks for the elaborate testing and level of details!

**Comment by cha0sCat** [comment]
works great this time
[output.webm](https://github.com/pouriyajamshidi/flat/assets/47235156/e260854a-95e5-49ba-9672-be69aae38dcf)

**Comment by pouriyajamshidi** [comment]
WOW! you got some traffic on that machine :D

Thanks a lot for your report, debugging and suggestions mate. Was very fun!

Btw, if you want, you can modify your PR to include the changes in the branch that fixed this for you. It might sounds a bit silly but since you did a lot of work on this, I think it'd be nice to have your name as a contributor of this project.

**Comment by cha0sCat** [comment]
Yes, I've cherrypicked your commit and cleaned up mine.

I also noticed that the portable version already has a code for initialize:
https://github.com/pouriyajamshidi/flat/blob/8dfba7e197e4474c8c61f6c7f8439686f094b09d/bpf/flat_portable.c#L142

**Comment by pouriyajamshidi** [comment]
The portable version uses the `perfevent` instead of `ringbuf` since `ringbuf` is not available on older Linux versions.

On the "non-portable" version we cannot zero-initialize `pkt` since `ringbuf` would not accept it :S

**Comment by cha0sCat** [comment]
😂 cool

**Comment by pouriyajamshidi** [comment]
fixed in #31
